### PR TITLE
Mi22007

### PR DESCRIPTION
**Pull request by MarkoCvijetinovic**

### CODE REVIEW

File: PROJECT-DESCRIPTION-TEMPLATE.md [hunk 1]
```diff
@@ -1,32 +1,36 @@
...
 ### Fundamental:
 
-[ ] Model with lighting
-[ ] Two types of lighting with customizable colors and movement through GUI or ACTIONS
-[ ] {ACTION_X} --- AFTER_M_SECONDS---Triggers---> {EVENT_A} ---> AFTER_N_SECONDS---Triggers---> {EVENT_B}
+[x] Model with lighting
+
+[x] Two types of lighting with customizable colors and movement through GUI or ACTIONS
+
+[x] {ACTION_X} --- AFTER_M_SECONDS---Triggers---> {EVENT_A} ---> AFTER_N_SECONDS---Triggers---> {EVENT_B}
```
Comment (spaske00) [inline PROJECT-DESCRIPTION-TEMPLATE.md]: Zameniti {} sa komandama i dešavanjima na sceni.

File: app/resources/shaders/planet.glsl
Comment (spaske00) [inline app/resources/shaders/planet.glsl:1]: [REFACTOR] planet.glsl deluje slično asteriod.glsl šejderu, da li mogu biti jedan šejder?

File: app/src/MainController.cpp [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    auto shaderBloom = resources->shader("bloom");
+
+    engine::graphics::OpenGL::initialize_bloom(platform->window()->width(), platform->window()->height(), shaderBlur,
+                                               shaderBloom);
+}
+
+void MainController::initialize_asteroids() {
+    m_modelMatrices = new glm::mat4[m_amount];
+    auto platform   = get<engine::platform::PlatformController>();
+    srand(static_cast<unsigned int>(platform->frame_time().current)); // initialize random seed
```
Comment (spaske00) [inline app/src/MainController.cpp]: [REFACTOR] Obrisati, nema potrebe za inicijalizacijom ako se koristi <random>

File: app/src/MainController.cpp [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+    auto shaderBlur  = resources->shader("blur");
+    auto shaderBloom = resources->shader("bloom");
+
+    engine::graphics::OpenGL::initialize_bloom(platform->window()->width(), platform->window()->height(), shaderBlur,
+                                               shaderBloom);
+}
+
+void MainController::initialize_asteroids() {
+    m_modelMatrices = new glm::mat4[m_amount];
```
Comment (spaske00) [inline app/src/MainController.cpp]: [REFACTOR] Koristiti std::vector u ovom slučaju umesto niza. Automatski će biti oslobođen na kraju programa.

File: app/src/MainController.cpp [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        model        = translate(model, m_csillaPos);
+
+        static std::random_device rd;
+        static std::mt19937 gen(rd());
+        static std::uniform_real_distribution dis_offset(-offset, offset);
+        static std::uniform_real_distribution dis_scale(0.003f, 0.015f);
+        static std::uniform_real_distribution dis_rotation(0.0f, 360.0f);
+
+        // 1. Translation: displace along a circle with 'radius' in range [-offset, offset]
+        float angle        = static_cast<float>(i) / static_cast<float>(m_amount) * 360.0f;
```
Comment (spaske00) [inline app/src/MainController.cpp]: [REFACTOR] Nema potrebe za static_cast u ovom slučaju, kompajler će uraditi automatsku konverziju u float dokle god je barem jedan element izraza float.

File: app/src/MainController.cpp [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    }
+}
+void MainController::alter_star() {
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    m_starLuminocity *= 1.5f;
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    m_starLuminocity /= 1.5f;
+    m_terranScale *= 2.0f;
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    m_terranScale /= 2.0f;
```
Comment (spaske00) [inline app/src/MainController.cpp]: [CRITICAL] Uh, nikako sleep, ovo će blokirati ceo program. Predlažem da postavite da se prilikom pritiska T intenzitet svetlosti zvezde postepeno povećava do nekog gornjeg ograničenja, onda bude nekoliko sekundi na tom gornjem ograničenju, i opet se postepeno smanji na početnu vrednost, postepeno.

File: engine/include/engine/resources/Mesh.hpp [hunk 6]
```diff
@@ -44,6 +44,18 @@ class Mesh {
...
 
+    [[nodiscard]] uint32_t VAO() const {
+        return m_vao;
+    }
+
+    [[nodiscard]] uint32_t num_of_indices() const {
+        return m_num_indices;
+    }
+
+    [[nodiscard]] std::vector<Texture *> textures() const {
```
Comment (spaske00) [inline engine/include/engine/resources/Mesh.hpp]: [REFACTOR] Vratiti `const std::vector<Texture *>&`, u suprotnom se pravi kopija vektora.

File: engine/src/OpenGL.cpp
Comment (spaske00) [inline engine/src/OpenGL.cpp:1]: [REFACTOR] Dodati CHECKED_GL_CALL za sve opengl pozive.

**Review by spaske00 — CHANGES_REQUESTED** [review]
Scena lepo izgleda, ostalo je doraditi par stavki, ostavio sam komentare. 

Datoteku engine/cmake-build-debug možete obrisati.

**Comment by MarkoCvijetinovic** [comment]
"[CRITICAL] Uh, nikako sleep, ovo će blokirati ceo program. Predlažem
da postavite da se prilikom pritiska T intenzitet svetlosti zvezde
postepeno povećava do nekog gornjeg ograničenja, onda bude nekoliko
sekundi na tom gornjem ograničenju, i opet se postepeno smanji na
početnu vrednost, postepeno."


Ne shvatam u čemu je problem sa sleep, ne zaustavlja se ceo program već
samo ta nit.
"[REFACTOR] planet.glsl deluje slično asteriod.glsl šejderu, da li mogu
biti jedan šejder?"

Razlikuju se u aInstanceMatrix. Mogao bih da deaktiviram pa ponovo da
aktiviram taj atribut, ali ima i problem u tome što će biti postavljeno na
(0.0, 0.0, 0.0, 1.0) i moram da menjam, to sve meni zvuči komplikovanije.

Ostalo sam promenio.

File: app/src/MainController.cpp (hunk 5 above)
Comment (spaske00) [reply app/src/MainController.cpp]: [CRITICAL]
> Ne shvatam u čemu je problem sa sleep, ne zaustavlja se ceo program već
samo ta nit.

U redu, promaklo mi je da se pravi nova nit, to može da radi, ali onda morate rešiti problem stvaranja nove niti na svaki pritisak dugmenta T i nedostatak sinhronizacije vrednosti m_starLuminocity.

**Comment by MarkoCvijetinovic** [comment]
Rešio sam probleme ograničivši na jednu nit, takođe stavio sam da se
postepeno menja kao što se predložili jer lepše izgleda.

File: app/src/MainController.cpp [hunk 7]
```diff
@@ -397,20 +395,21 @@ void MainController::poll_events() {
         m_spotLightColor[2] = std::max(m_spotLightColor[2] - 0.02f, 0.0f);
     }
 
-    if (platform->key(engine::platform::KeyId::KEY_T).is_down() && !m_starKeyPressed) {
+    if (platform->key(engine::platform::KeyId::KEY_T).is_down() && !m_starEnabled) {
```
Comment (spaske00) [inline app/src/MainController.cpp]: Ovaj kod i dalje nije `thread-safe`, naime:
glavna nit može pročitati `m_starEnabled = false` i ući u if-granu, i startovati nit sa `std::thread(&MainController::alter_star, this).detach();` i odmah nakon te naredbe biti sklonjena sa CPU-a od strane kernela, zatim novo napravljena nit se može izvršiti do kraja, i onda se nastaviti izvršavanje glavne niti koja će postaviti `m_StarEnabled=true` i nakon toga okidanje ovog događaja više neće biti moguće. 

prilikom provere m_starEnabled koristite `compare_exchange` i onda neka nova nit samo vrati vrednost kada završi,
`m_starEnabled=true` na liniji 400 je onda višak

**Comment by MarkoCvijetinovic** [comment]
Hvala na povratnoj informaciji. Ispravio sam.

**Comment by MarkoCvijetinovic** [comment]
Da li je sada sve okej?

**Comment by spaske00** [comment]
Postoji još delova koda koji bi mogli dodatno da se utegnu, ali sve u svemu projekat izgleda dobro.

**Comment by MarkoCvijetinovic** [comment]
Koliko sam poena ostvario za sad?

**Comment by MarkoCvijetinovic** [comment]
Samo da Vas podsetim da mi još niste odgovorili koliko poena imam na
projektu.

**Comment by spaske00** [comment]
Bloom ne radi, slika u prilogu. 
Na ostatku projekta imate 25. Molim Vas, popravite bloom kako bih mogao da Vam dam maksimalnih 35.
bloom

**Comment by MarkoCvijetinovic** [comment]
Kod mene izgleda normalno, a sinhronizovano mi je sa origin-om.

**Comment by spaske00** [comment]
Dešava se na sveže kloniranom projektu, možete li da mi pošaljete snimak ekrana sa bloom on/off?

**Comment by MarkoCvijetinovic** [comment]
Moj drug je maločas klonirao moj projekat i radi mu.
 Screencast from 2025-08-18 17-42-51.webm
<https://drive.google.com/file/d/1p16yBsnb7XjXAWURklRFYXB_wagd90RH/view?usp=drive_web>

**Comment by spaske00** [comment]
Baš čudno, sve u redu, 35. Pogledaću o čemu se radi prvom prilikom.
